In [1]:
# Imports
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import numbers

from pandas.core.dtypes.common import is_numeric_dtype

### Dataset Source
Link: https://www.kaggle.com/datasets/adilshamim8/economic-indicators-and-inflation/

In [2]:
# Load Datasets
economic_indicators_and_inflation = pd.read_csv('../data/task03/Economic Indicators And Inflation.csv')
development_indicators = pd.read_excel('../data/task03/WorldBank.xlsx')

print(f'Economic Indicators and Inflation:\n{economic_indicators_and_inflation}')
print('----------------------------------')
print(f'Development Indicators:\n{development_indicators}')

Economic Indicators and Inflation:
      Country   Year   GDP (in billion USD)   Inflation Rate (%)  \
0         USA   2010                15000.0                 1.64   
1         USA   2011                15500.0                 3.16   
2         USA   2012                16000.0                 2.07   
3         USA   2013                16500.0                 1.50   
4         USA   2014                17000.0                 1.62   
..        ...    ...                    ...                  ...   
299  Pakistan   2021                  296.0                 8.90   
300  Pakistan   2022                  350.0                12.30   
301  Pakistan   2023                  400.0                20.00   
302  Pakistan   2024                  450.0                25.00   
303  Pakistan   2025                  500.0                20.00   

      Unemployment Rate (%)   Economic Growth (%)  
0                      9.63                  2.55  
1                      8.94                 

In [3]:
# Functions

def headers_to_snake_case(df):
    """Renames headers to snake_case"""
    df = df.copy()

    df.columns = (
        df.columns.astype(str)
        .str.strip()
        .str.lower()
        .str.replace(r"[^a-z0-9]+", "_", regex=True)
        .str.strip('_')
    )
    return df


def check_values_below_zero(df):
    """Checks if there are any values below zero"""
    for col_name in df.columns:
        if is_numeric_dtype(df[col_name]):
            invalid_value = df.loc[
                (df[col_name] < 0)
            ]
            print(f'Below zero values in "{col_name}": {len(invalid_value)}')

    return f'All Columns checked'


def add_categories(
        df,
        col,
        category_col_name,
        bins,
        labels,
        right=True,
        retbins=False,
        precision=3,
        include_lowest=False,
        duplicates='raise',
        ordered=True
):
    """Creates categories with pd.cut()"""
    df[category_col_name] = pd.cut(
        df[col],
        bins=bins,
        labels=labels,
        right=right,
        retbins=retbins,
        precision=precision,
        include_lowest=include_lowest,
        duplicates=duplicates,
        ordered=ordered
    )
    return df


def get_q1_q3(df, col):
    q1 = df[col].quantile(0.25)
    q3 = df[col].quantile(0.75)
    return q1, q3


def highest(df, by_col):
    """Gets the highest value"""
    try:
        return df.loc[df[by_col].idxmax()]
    except KeyError:
        print(f"No column named {by_col}")
        return None


def lowest(df, by_col):
    """Gets the lowest value"""
    try:
        return df.loc[df[by_col].idxmin()]
    except KeyError:
        print(f"No column named {by_col}")


def format_value(value):
    """Format value if it is a number"""
    if isinstance(value, numbers.Number):
        return f"{value:.2f}"
    else:
        return str(value)


def highest_print(rank_type, name, rank_value):
    return (f'Highest average {rank_type.lower()}:'
            f'\n- {name}'
            f'\n- {rank_type}: {format_value(rank_value)}')


def lowest_print(rank_type, name, rank_value):
    return (f'Lowest average {rank_type.lower()}:'
            f'\n- {name}'
            f'\n- {rank_type}: {format_value(rank_value)}')


In [4]:
# Shape
print(f'Economic Indicators and Inflation: {economic_indicators_and_inflation.shape}')
print(f'Development Indicators: {development_indicators.shape}')

Economic Indicators and Inflation: (304, 6)
Development Indicators: (12449, 15)


In [5]:
# Column Headers

print(f'Economic Indicators and Inflation Headers: {', '.join(economic_indicators_and_inflation.columns)}')
print()
print(f'Development Indicators Headers: {', '.join(development_indicators.columns)}')

Economic Indicators and Inflation Headers: Country,  Year,  GDP (in billion USD),  Inflation Rate (%),  Unemployment Rate (%),  Economic Growth (%)

Development Indicators Headers: Country Name, Country Code, Region, IncomeGroup, Year, Birth rate, crude (per 1,000 people), Death rate, crude (per 1,000 people), Electric power consumption (kWh per capita), GDP (USD), GDP per capita (USD), Individuals using the Internet (% of population), Infant mortality rate (per 1,000 live births), Life expectancy at birth (years), Population density (people per sq. km of land area), Unemployment (% of total labor force) (modeled ILO estimate)


### Possible Similar columns
Economic Indicators and Inflation  - Development Indicators
- Country and Country Name
- Year and Year
- Unemployment Rate (%) and Unemployment (% of total labor force) (modeled ILO estimate)

###### ***This will be analyzed further before merging the tables***

#### Inspect Datasets

In [6]:
# Economic Indicators and Inflation Dataset Info
economic_indicators_and_inflation.info()

<class 'pandas.DataFrame'>
RangeIndex: 304 entries, 0 to 303
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Country                 304 non-null    str    
 1    Year                   304 non-null    int64  
 2    GDP (in billion USD)   304 non-null    float64
 3    Inflation Rate (%)     304 non-null    float64
 4    Unemployment Rate (%)  304 non-null    float64
 5    Economic Growth (%)    304 non-null    float64
dtypes: float64(4), int64(1), str(1)
memory usage: 14.4 KB


In [7]:
# Economic Indicators and Inflation Dataset Header Titles
economic_indicators_and_inflation.columns

Index(['Country', ' Year', ' GDP (in billion USD)', ' Inflation Rate (%)',
       ' Unemployment Rate (%)', ' Economic Growth (%)'],
      dtype='str')

In [8]:
# Rename Headers of Economic Indicators and Inflation
economic_indicators_and_inflation = headers_to_snake_case(economic_indicators_and_inflation)
economic_indicators_and_inflation.columns

Index(['country', 'year', 'gdp_in_billion_usd', 'inflation_rate',
       'unemployment_rate', 'economic_growth'],
      dtype='str')

In [9]:
# Checking for Null Values in Economic Indicators and Inflation Dataset
economic_indicators_and_inflation.isna().sum()

country               0
year                  0
gdp_in_billion_usd    0
inflation_rate        0
unemployment_rate     0
economic_growth       0
dtype: int64

In [10]:
# Development Indicators Dataset Info
development_indicators.info()

<class 'pandas.DataFrame'>
RangeIndex: 12449 entries, 0 to 12448
Data columns (total 15 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   Country Name                                                  12449 non-null  str    
 1   Country Code                                                  12449 non-null  str    
 2   Region                                                        12449 non-null  str    
 3   IncomeGroup                                                   12449 non-null  str    
 4   Year                                                          12449 non-null  int64  
 5   Birth rate, crude (per 1,000 people)                          11440 non-null  float64
 6   Death rate, crude (per 1,000 people)                          11416 non-null  float64
 7   Electric power consumption (kWh per capita)                   5848 non-null   

In [11]:
# Development Indicators Dataset Header Titles
development_indicators.columns

Index(['Country Name', 'Country Code', 'Region', 'IncomeGroup', 'Year',
       'Birth rate, crude (per 1,000 people)',
       'Death rate, crude (per 1,000 people)',
       'Electric power consumption (kWh per capita)', 'GDP (USD)',
       'GDP per capita (USD)',
       'Individuals using the Internet (% of population)',
       'Infant mortality rate (per 1,000 live births)',
       'Life expectancy at birth (years)',
       'Population density (people per sq. km of land area)',
       'Unemployment (% of total labor force) (modeled ILO estimate)'],
      dtype='str')

In [12]:
# Rename Development Indicators Headers
development_indicators = headers_to_snake_case(development_indicators)
development_indicators.columns

Index(['country_name', 'country_code', 'region', 'incomegroup', 'year',
       'birth_rate_crude_per_1_000_people',
       'death_rate_crude_per_1_000_people',
       'electric_power_consumption_kwh_per_capita', 'gdp_usd',
       'gdp_per_capita_usd', 'individuals_using_the_internet_of_population',
       'infant_mortality_rate_per_1_000_live_births',
       'life_expectancy_at_birth_years',
       'population_density_people_per_sq_km_of_land_area',
       'unemployment_of_total_labor_force_modeled_ilo_estimate'],
      dtype='str')

In [13]:
development_indicators = development_indicators.rename(columns={'incomegroup': 'income_group'})
development_indicators.columns

Index(['country_name', 'country_code', 'region', 'income_group', 'year',
       'birth_rate_crude_per_1_000_people',
       'death_rate_crude_per_1_000_people',
       'electric_power_consumption_kwh_per_capita', 'gdp_usd',
       'gdp_per_capita_usd', 'individuals_using_the_internet_of_population',
       'infant_mortality_rate_per_1_000_live_births',
       'life_expectancy_at_birth_years',
       'population_density_people_per_sq_km_of_land_area',
       'unemployment_of_total_labor_force_modeled_ilo_estimate'],
      dtype='str')

In [14]:
# Checking for Null Values in Development Indicators
development_indicators.isna().sum().head(20)

country_name                                                 0
country_code                                                 0
region                                                       0
income_group                                                 0
year                                                         0
birth_rate_crude_per_1_000_people                         1009
death_rate_crude_per_1_000_people                         1033
electric_power_consumption_kwh_per_capita                 6601
gdp_usd                                                   2871
gdp_per_capita_usd                                        2874
individuals_using_the_internet_of_population              7385
infant_mortality_rate_per_1_000_live_births               2465
life_expectancy_at_birth_years                            1273
population_density_people_per_sq_km_of_land_area           604
unemployment_of_total_labor_force_modeled_ilo_estimate    7241
dtype: int64

In [15]:
print(
    f'Economic Indicators and Inflation \n '
    f'min year: {economic_indicators_and_inflation.year.min()}\n '
    f'max year: {economic_indicators_and_inflation.year.max()}'
)
print()
print(
    f'Development Indicators \n '
    f'min year: {development_indicators.year.min()}\n '
    f'max year: {development_indicators.year.max()}'
)

Economic Indicators and Inflation 
 min year: 2010
 max year: 2025

Development Indicators 
 min year: 1960
 max year: 2018


#### The years in common are from 2010 to 2018
I will reduce both datasets to the years in common

In [16]:
econom_ind_and_inflation_2010_2018 = economic_indicators_and_inflation.loc[
    economic_indicators_and_inflation['year'].between(2010, 2018)
]

dev_ind_2010_2018 = development_indicators.loc[
    development_indicators['year'].between(2010, 2018)
]

print(econom_ind_and_inflation_2010_2018)
print()
print(dev_ind_2010_2018)

      country  year  gdp_in_billion_usd  inflation_rate  unemployment_rate  \
0         USA  2010             15000.0            1.64               9.63   
1         USA  2011             15500.0            3.16               8.94   
2         USA  2012             16000.0            2.07               8.10   
3         USA  2013             16500.0            1.50               7.70   
4         USA  2014             17000.0            1.62               7.25   
..        ...   ...                 ...             ...                ...   
292  Pakistan  2014               247.0            8.60               6.00   
293  Pakistan  2015               271.0            4.50               5.50   
294  Pakistan  2016               278.0            3.80               5.00   
295  Pakistan  2017               314.0            4.00               5.00   
296  Pakistan  2018               314.0            3.90               5.00   

     economic_growth  
0               2.55  
1               1

In [17]:
# NaN values in Development Indicators after reducing dataframe
dev_ind_2010_2018.isna().sum()

country_name                                                 0
country_code                                                 0
region                                                       0
income_group                                                 0
year                                                         0
birth_rate_crude_per_1_000_people                          284
death_rate_crude_per_1_000_people                          284
electric_power_consumption_kwh_per_capita                 1200
gdp_usd                                                    113
gdp_per_capita_usd                                         113
individuals_using_the_internet_of_population               227
infant_mortality_rate_per_1_000_live_births                189
life_expectancy_at_birth_years                             323
population_density_people_per_sq_km_of_land_area            29
unemployment_of_total_labor_force_modeled_ilo_estimate     225
dtype: int64

The World Bank development indicators contain missing values because not all countries report all indicators for every year. Since the dataset contains many indicators, removing all rows with missing values would result in unnecessary data loss. Therefore, missing values will be  handled selectively. Rows will be not removed globally. Instead, missing values will be kept as NaN, and calculations will be performed using available observations. For analyses requiring specific indicators, only rows with missing values in those specific columns will be excluded. Columns with very high missingness, such as electric power consumption, are kept but will be not used as core indicators in the main analysis.

In [18]:
# Check percentage of missing values
missing_summary = (
    dev_ind_2010_2018
    .isna()
    .sum()
    .to_frame('missing_count')
)

missing_summary['missing_percent'] = (
    missing_summary['missing_count'] / len(dev_ind_2010_2018) * 100
)

print(missing_summary.sort_values('missing_percent', ascending=False))

                                                    missing_count  \
electric_power_consumption_kwh_per_capita                    1200   
life_expectancy_at_birth_years                                323   
birth_rate_crude_per_1_000_people                             284   
death_rate_crude_per_1_000_people                             284   
individuals_using_the_internet_of_population                  227   
unemployment_of_total_labor_force_modeled_ilo_e...            225   
infant_mortality_rate_per_1_000_live_births                   189   
gdp_usd                                                       113   
gdp_per_capita_usd                                            113   
population_density_people_per_sq_km_of_land_area               29   
country_name                                                    0   
country_code                                                    0   
region                                                          0   
income_group                      

In [19]:
# Important and optional columns
core_dev_ind_cols = [
    "country_name",
    "country_code",
    "region",
    "income_group",
    "year",
    "gdp_usd",
    "gdp_per_capita_usd",
    "individuals_using_the_internet_of_population",
    "life_expectancy_at_birth_years",
    "population_density_people_per_sq_km_of_land_area",
    "unemployment_of_total_labor_force_modeled_ilo_estimate",
    "birth_rate_crude_per_1_000_people",
    "death_rate_crude_per_1_000_people"
]

optional_dev_ind_cols = [
    "electric_power_consumption_kwh_per_capita",
    "infant_mortality_rate_per_1_000_live_births"
]

In [20]:
# Development Indicators core table
dev_ind_2010_2018_core = dev_ind_2010_2018[core_dev_ind_cols].copy()
dev_ind_2010_2018_core

,country_name,country_code,region,income_group,year,gdp_usd,gdp_per_capita_usd,individuals_using_the_internet_of_population,life_expectancy_at_birth_years,population_density_people_per_sq_km_of_land_area,unemployment_of_total_labor_force_modeled_ilo_estimate,birth_rate_crude_per_1_000_people,death_rate_crude_per_1_000_people
0,Afghanistan,AFG,South Asia,Low income,2018,1.936300e+10,520.897,NaN,NaN,56.9378,1.542,NaN,NaN
1,Afghanistan,AFG,South Asia,Low income,2017,2.019180e+10,556.302,13.5000,64.130,55.5960,1.559,33.211,6.575
2,Afghanistan,AFG,South Asia,Low income,2016,1.936260e+10,547.228,11.2000,63.763,54.1971,1.634,33.981,6.742
3,Afghanistan,AFG,South Asia,Low income,2015,1.990710e+10,578.466,8.2600,63.377,52.7121,1.679,34.809,6.929
4,Afghanistan,AFG,South Asia,Low income,2014,2.048490e+10,613.856,7.0000,62.966,51.1148,1.735,35.706,7.141
...,...,...,...,...,...,...,...,...,...,...,...,...,...
12394,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2014,1.949550e+10,1434.900,16.3647,58.410,35.1213,5.518,34.978,9.277
12395,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2013,1.909100e+10,1430.000,15.5000,56.897,34.5104,5.623,35.768,10.115
12396,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2012,1.711480e+10,1304.970,12.0000,55.032,33.9024,5.608,36.292,11.167
12397,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2011,1.410190e+10,1093.650,8.4000,52.896,33.3316,5.368,36.526,12.384


In [21]:
# Duplicate rows
print(
    f'Economic Indicators and Inflation duplicated rows: '
      f'{econom_ind_and_inflation_2010_2018.duplicated().sum()}'
)

print(f'Development Indicators duplicated rows: {dev_ind_2010_2018_core.duplicated().sum()}')


Economic Indicators and Inflation duplicated rows: 0
Development Indicators duplicated rows: 0


In [22]:
econom_ind_and_inflation_2010_2018.info()

<class 'pandas.DataFrame'>
Index: 171 entries, 0 to 296
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             171 non-null    str    
 1   year                171 non-null    int64  
 2   gdp_in_billion_usd  171 non-null    float64
 3   inflation_rate      171 non-null    float64
 4   unemployment_rate   171 non-null    float64
 5   economic_growth     171 non-null    float64
dtypes: float64(4), int64(1), str(1)
memory usage: 9.4 KB


In [23]:
# Values bellow zero
dataframes = [econom_ind_and_inflation_2010_2018, dev_ind_2010_2018_core]
for df in dataframes:
    print(check_values_below_zero(df))

Below zero values in "year": 0
Below zero values in "gdp_in_billion_usd": 0
Below zero values in "inflation_rate": 4
Below zero values in "unemployment_rate": 0
Below zero values in "economic_growth": 10
All Columns checked
Below zero values in "year": 0
Below zero values in "gdp_usd": 0
Below zero values in "gdp_per_capita_usd": 0
Below zero values in "individuals_using_the_internet_of_population": 0
Below zero values in "life_expectancy_at_birth_years": 0
Below zero values in "population_density_people_per_sq_km_of_land_area": 0
Below zero values in "unemployment_of_total_labor_force_modeled_ilo_estimate": 0
Below zero values in "birth_rate_crude_per_1_000_people": 0
Below zero values in "death_rate_crude_per_1_000_people": 0
All Columns checked


Two columns contain values below zero: inflation_rate and economic_growth. Both are part of the Economic Indicators and Inflation dataset. Negative inflation rates are economically meaningful because they represent deflation, while negative economic growth values indicate economic contraction. Therefore, these values were not removed during the cleaning process.

In [24]:
# Potential outliers
economic_indicators_and_inflation_outliers = econom_ind_and_inflation_2010_2018.loc[
    (econom_ind_and_inflation_2010_2018['inflation_rate'] < -10) |
    (econom_ind_and_inflation_2010_2018['inflation_rate'] > 20)  |
    (econom_ind_and_inflation_2010_2018['economic_growth'] < -10) |
    (econom_ind_and_inflation_2010_2018['economic_growth'] > 20)
]

economic_indicators_and_inflation_outliers

,country,year,gdp_in_billion_usd,inflation_rate,unemployment_rate,economic_growth
181,Saudi Arabia,2015,646.0,5.4,10.5,-14.0
242,Indonesia,2012,1.0,84.0,4.3,6.1
243,Indonesia,2013,1.0,333.0,6.4,5.9
244,Indonesia,2014,1.0,492.0,8.4,5.9
245,Indonesia,2015,1.0,22.0,6.4,6.2
246,Indonesia,2016,1.0,34.0,3.5,5.6
247,Indonesia,2017,1.0,40.0,3.6,5.5
248,Indonesia,2018,1.0,42.0,3.2,5.3
257,Turkey,2011,1.0,226.0,10.4,9.8
258,Turkey,2012,1.0,384.0,6.2,8.4


In [25]:
dev_ind_outliers = dev_ind_2010_2018_core.loc[
    (dev_ind_2010_2018_core["individuals_using_the_internet_of_population"] < 0) |
    (dev_ind_2010_2018_core["individuals_using_the_internet_of_population"] > 100) |

    (dev_ind_2010_2018_core["life_expectancy_at_birth_years"] < 40) |
    (dev_ind_2010_2018_core["life_expectancy_at_birth_years"] > 90) |

    (dev_ind_2010_2018_core["population_density_people_per_sq_km_of_land_area"] < 0) |
    (dev_ind_2010_2018_core["population_density_people_per_sq_km_of_land_area"] > 10000) |

    (dev_ind_2010_2018_core["unemployment_of_total_labor_force_modeled_ilo_estimate"] < 0) |
    (dev_ind_2010_2018_core["unemployment_of_total_labor_force_modeled_ilo_estimate"] > 100) |

    (dev_ind_2010_2018_core["birth_rate_crude_per_1_000_people"] < 0) |
    (dev_ind_2010_2018_core["birth_rate_crude_per_1_000_people"] > 60) |

    (dev_ind_2010_2018_core["death_rate_crude_per_1_000_people"] < 0) |
    (dev_ind_2010_2018_core["death_rate_crude_per_1_000_people"] > 30)
]

dev_ind_outliers

,country_name,country_code,region,income_group,year,gdp_usd,gdp_per_capita_usd,individuals_using_the_internet_of_population,life_expectancy_at_birth_years,population_density_people_per_sq_km_of_land_area,unemployment_of_total_labor_force_modeled_ilo_estimate,birth_rate_crude_per_1_000_people,death_rate_crude_per_1_000_people
6608,"Macao SAR, China",MAC,East Asia & Pacific,High income: nonOECD,2018,5.454520e+10,86355.4,83.7941,NaN,20777.5,2.023,NaN,NaN
6609,"Macao SAR, China",MAC,East Asia & Pacific,High income: nonOECD,2017,5.055940e+10,81208.9,83.1741,83.989,20479.8,2.000,11.200,3.866
6610,"Macao SAR, China",MAC,East Asia & Pacific,High income: nonOECD,2016,4.532230e+10,73955.0,81.6430,83.854,20159.1,1.900,11.310,3.800
6611,"Macao SAR, China",MAC,East Asia & Pacific,High income: nonOECD,2015,4.536170e+10,75341.0,77.6001,83.707,19805.4,1.810,11.360,3.743
6612,"Macao SAR, China",MAC,East Asia & Pacific,High income: nonOECD,2014,5.534800e+10,93777.1,69.7800,83.543,19478.8,1.700,11.317,3.700
6613,"Macao SAR, China",MAC,East Asia & Pacific,High income: nonOECD,2013,5.155210e+10,89287.5,65.8000,83.361,19055.2,1.820,11.165,3.671
6614,"Macao SAR, China",MAC,East Asia & Pacific,High income: nonOECD,2012,4.303160e+10,76291.8,61.3100,83.158,18864.2,1.970,10.904,3.657
6615,"Macao SAR, China",MAC,East Asia & Pacific,High income: nonOECD,2011,3.670990e+10,66644.4,60.2037,82.938,18422.5,2.560,10.544,3.657
6616,"Macao SAR, China",MAC,East Asia & Pacific,High income: nonOECD,2010,2.812360e+10,52253.2,55.1980,82.704,18121.9,2.840,10.100,3.668
7317,Monaco,MCO,Europe & Central Asia,High income: nonOECD,2017,6.400950e+09,166726.0,97.0530,NaN,19196.0,NaN,NaN,NaN


Both datasets contain potential outliers. In the Economic Indicators and Inflation dataset, potential outliers were identified in the inflation_rate and economic_growth columns. In the Development Indicators dataset, potential outliers were identified in the population_density_people_per_sq_km_of_land_area column. These values were not removed automatically, because extreme values may be valid for specific countries and years. Instead, they were flagged for further review.

In [26]:
# Convert GDP from billion USD to USD in the Economic Indicators dataset
econom_ind_and_inflation_2010_2018['gdb_usd_economic'] = (
    econom_ind_and_inflation_2010_2018['gdp_in_billion_usd'] * 1_000_000_000
)

econom_ind_and_inflation_2010_2018

,country,year,gdp_in_billion_usd,inflation_rate,unemployment_rate,economic_growth,gdb_usd_economic
0,USA,2010,15000.0,1.64,9.63,2.55,1.500000e+13
1,USA,2011,15500.0,3.16,8.94,1.53,1.550000e+13
2,USA,2012,16000.0,2.07,8.10,2.28,1.600000e+13
3,USA,2013,16500.0,1.50,7.70,1.84,1.650000e+13
4,USA,2014,17000.0,1.62,7.25,2.53,1.700000e+13
...,...,...,...,...,...,...,...
292,Pakistan,2014,247.0,8.60,6.00,4.10,2.470000e+11
293,Pakistan,2015,271.0,4.50,5.50,4.70,2.710000e+11
294,Pakistan,2016,278.0,3.80,5.00,5.40,2.780000e+11
295,Pakistan,2017,314.0,4.00,5.00,5.40,3.140000e+11


In [27]:
# Rename Development Indicator GDP column for clarity when merging both tables in the future
dev_ind_2010_2018_core = dev_ind_2010_2018_core.rename(
    columns={
        'gdp_usd': 'gdb_usd_worldbank',
    }
)

dev_ind_2010_2018_core

,country_name,country_code,region,income_group,year,gdb_usd_worldbank,gdp_per_capita_usd,individuals_using_the_internet_of_population,life_expectancy_at_birth_years,population_density_people_per_sq_km_of_land_area,unemployment_of_total_labor_force_modeled_ilo_estimate,birth_rate_crude_per_1_000_people,death_rate_crude_per_1_000_people
0,Afghanistan,AFG,South Asia,Low income,2018,1.936300e+10,520.897,NaN,NaN,56.9378,1.542,NaN,NaN
1,Afghanistan,AFG,South Asia,Low income,2017,2.019180e+10,556.302,13.5000,64.130,55.5960,1.559,33.211,6.575
2,Afghanistan,AFG,South Asia,Low income,2016,1.936260e+10,547.228,11.2000,63.763,54.1971,1.634,33.981,6.742
3,Afghanistan,AFG,South Asia,Low income,2015,1.990710e+10,578.466,8.2600,63.377,52.7121,1.679,34.809,6.929
4,Afghanistan,AFG,South Asia,Low income,2014,2.048490e+10,613.856,7.0000,62.966,51.1148,1.735,35.706,7.141
...,...,...,...,...,...,...,...,...,...,...,...,...,...
12394,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2014,1.949550e+10,1434.900,16.3647,58.410,35.1213,5.518,34.978,9.277
12395,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2013,1.909100e+10,1430.000,15.5000,56.897,34.5104,5.623,35.768,10.115
12396,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2012,1.711480e+10,1304.970,12.0000,55.032,33.9024,5.608,36.292,11.167
12397,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2011,1.410190e+10,1093.650,8.4000,52.896,33.3316,5.368,36.526,12.384


#### Merge Datasets

In [28]:
# Merge check

merge_check = pd.merge(
    econom_ind_and_inflation_2010_2018,
    dev_ind_2010_2018_core,
    left_on=['country', 'year'],
    right_on=['country_name', 'year'],
    how='outer',
    indicator=True
)

merge_check['_merge'].value_counts()

_merge
right_only    1764
both           135
left_only       36
Name: count, dtype: int64

In [29]:
merge_check.loc[
    merge_check['_merge'] == 'left_only',
].sort_values(['country', 'year'], ascending=[True, False])

,country,year,gdp_in_billion_usd,inflation_rate,unemployment_rate,economic_growth,gdb_usd_economic,country_name,country_code,region,income_group,gdb_usd_worldbank,gdp_per_capita_usd,individuals_using_the_internet_of_population,life_expectancy_at_birth_years,population_density_people_per_sq_km_of_land_area,unemployment_of_total_labor_force_modeled_ilo_estimate,birth_rate_crude_per_1_000_people,death_rate_crude_per_1_000_people,_merge
1403,Russia,2018,1700.0,2.90,4.80,2.30,1.700000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1402,Russia,2017,1650.0,3.70,5.20,1.50,1.650000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1401,Russia,2016,1600.0,5.40,5.60,-0.20,1.600000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1400,Russia,2015,1700.0,12.90,5.60,-3.70,1.700000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1399,Russia,2014,1800.0,15.50,5.60,0.50,1.800000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1398,Russia,2013,1750.0,6.80,5.20,1.80,1.750000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1397,Russia,2012,1700.0,6.60,5.40,3.40,1.700000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1396,Russia,2011,1600.0,6.10,6.10,4.30,1.600000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1395,Russia,2010,1500.0,8.80,7.50,4.00,1.500000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
1565,South Korea,2018,1400.0,1.50,3.80,2.70,1.400000e+12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only


In [30]:
merge_check.loc[
    merge_check['_merge'] == 'right_only',
].sort_values(['country_name', 'year'], ascending=[False, False])

,country,year,gdp_in_billion_usd,inflation_rate,unemployment_rate,economic_growth,gdb_usd_economic,country_name,country_code,region,income_group,gdb_usd_worldbank,gdp_per_capita_usd,individuals_using_the_internet_of_population,life_expectancy_at_birth_years,population_density_people_per_sq_km_of_land_area,unemployment_of_total_labor_force_modeled_ilo_estimate,birth_rate_crude_per_1_000_people,death_rate_crude_per_1_000_people,_merge
1934,NaN,2018,NaN,NaN,NaN,NaN,NaN,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,3.100050e+10,2147.000,NaN,NaN,37.3246,4.915,NaN,NaN,right_only
1933,NaN,2017,NaN,NaN,NaN,NaN,NaN,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2.281300e+10,1602.400,27.05550,60.812,36.8017,4.943,31.732,8.044,right_only
1932,NaN,2016,NaN,NaN,NaN,NaN,NaN,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,2.054870e+10,1464.580,23.12000,60.294,36.2683,5.239,32.864,8.286,right_only
1931,NaN,2015,NaN,NaN,NaN,NaN,NaN,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,1.996310e+10,1445.070,22.74280,59.534,35.7106,5.438,33.981,8.673,right_only
1930,NaN,2014,NaN,NaN,NaN,NaN,NaN,Zimbabwe,ZWE,Sub-Saharan Africa,Low income,1.949550e+10,1434.900,16.36470,58.410,35.1213,5.518,34.978,9.277,right_only
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4,NaN,2014,NaN,NaN,NaN,NaN,NaN,Afghanistan,AFG,South Asia,Low income,2.048490e+10,613.856,7.00000,62.966,51.1148,1.735,35.706,7.141,right_only
3,NaN,2013,NaN,NaN,NaN,NaN,NaN,Afghanistan,AFG,South Asia,Low income,2.056110e+10,637.165,5.90000,62.525,49.4280,1.725,36.670,7.380,right_only
2,NaN,2012,NaN,NaN,NaN,NaN,NaN,Afghanistan,AFG,South Asia,Low income,2.000160e+10,641.872,5.45455,62.054,47.7306,1.692,37.690,7.645,right_only
1,NaN,2011,NaN,NaN,NaN,NaN,NaN,Afghanistan,AFG,South Asia,Low income,1.780430e+10,591.162,5.00000,61.553,46.1315,1.984,38.750,7.936,right_only


In [31]:
# Country mapping
country_name_mapping = {
    "USA": "United States",
    "UK": "United Kingdom",
    "Russia": "Russian Federation",
    "South Korea": "Korea, Rep."
}

econom_ind_and_inflation_2010_2018['country'] = (
    econom_ind_and_inflation_2010_2018['country'].replace(country_name_mapping)
)

merge_check = pd.merge(
    econom_ind_and_inflation_2010_2018,
    dev_ind_2010_2018_core,
    left_on=["country", "year"],
    right_on=["country_name", "year"],
    how="outer",
    indicator=True
)

print(merge_check["_merge"].value_counts())



_merge
right_only    1728
both           171
left_only        0
Name: count, dtype: int64


After standardizing country names, the merge check showed 171 matched country-year records and 0 unmatched records from the Economic Indicators dataset. This means that all Economic Indicators records successfully matched with the World Bank dataset. The remaining 1728 right_only records belong only to the World Bank dataset, which contains many more countries.

In [32]:
economic_worldbank_merged = pd.merge(
    econom_ind_and_inflation_2010_2018,
    dev_ind_2010_2018_core,
    left_on=["country", "year"],
    right_on=["country_name", "year"],
    how='inner',
)

economic_worldbank_merged.shape


(171, 19)

In [33]:
economic_worldbank_merged.info()

<class 'pandas.DataFrame'>
RangeIndex: 171 entries, 0 to 170
Data columns (total 19 columns):
 #   Column                                                  Non-Null Count  Dtype  
---  ------                                                  --------------  -----  
 0   country                                                 171 non-null    str    
 1   year                                                    171 non-null    int64  
 2   gdp_in_billion_usd                                      171 non-null    float64
 3   inflation_rate                                          171 non-null    float64
 4   unemployment_rate                                       171 non-null    float64
 5   economic_growth                                         171 non-null    float64
 6   gdb_usd_economic                                        171 non-null    float64
 7   country_name                                            171 non-null    str    
 8   country_code                                       

In [34]:
# Fix typo

economic_worldbank_merged = economic_worldbank_merged.rename(
    columns={
        'gdb_usd_worldbank': 'gdp_usd_worldbank',
        'gdb_usd_economic': 'gdp_usd_economic',
    }
)

economic_worldbank_merged.info()

<class 'pandas.DataFrame'>
RangeIndex: 171 entries, 0 to 170
Data columns (total 19 columns):
 #   Column                                                  Non-Null Count  Dtype  
---  ------                                                  --------------  -----  
 0   country                                                 171 non-null    str    
 1   year                                                    171 non-null    int64  
 2   gdp_in_billion_usd                                      171 non-null    float64
 3   inflation_rate                                          171 non-null    float64
 4   unemployment_rate                                       171 non-null    float64
 5   economic_growth                                         171 non-null    float64
 6   gdp_usd_economic                                        171 non-null    float64
 7   country_name                                            171 non-null    str    
 8   country_code                                       

In [35]:
economic_worldbank_merged


,country,year,gdp_in_billion_usd,inflation_rate,unemployment_rate,economic_growth,gdp_usd_economic,country_name,country_code,region,income_group,gdp_usd_worldbank,gdp_per_capita_usd,individuals_using_the_internet_of_population,life_expectancy_at_birth_years,population_density_people_per_sq_km_of_land_area,unemployment_of_total_labor_force_modeled_ilo_estimate,birth_rate_crude_per_1_000_people,death_rate_crude_per_1_000_people
0,United States,2010,15000.0,1.64,9.63,2.55,1.500000e+13,United States,USA,North America,High income: OECD,1.500000e+13,48466.80,71.6900,78.5415,33.8157,9.633,13.000,7.995
1,United States,2011,15500.0,3.16,8.94,1.53,1.550000e+13,United States,USA,North America,High income: OECD,1.550000e+13,49883.10,69.7295,78.6415,34.0621,8.949,12.700,8.073
2,United States,2012,16000.0,2.07,8.10,2.28,1.600000e+13,United States,USA,North America,High income: OECD,1.620000e+13,51603.50,74.7000,78.7415,34.3129,8.069,12.600,8.102
3,United States,2013,16500.0,1.50,7.70,1.84,1.650000e+13,United States,USA,North America,High income: OECD,1.680000e+13,53106.90,71.4000,78.7415,34.5516,7.375,12.400,8.215
4,United States,2014,17000.0,1.62,7.25,2.53,1.700000e+13,United States,USA,North America,High income: OECD,1.750000e+13,55033.00,73.0000,78.8415,34.8061,6.168,12.500,8.237
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
166,Pakistan,2014,247.0,8.60,6.00,4.10,2.470000e+11,Pakistan,PAK,South Asia,Lower middle income,2.440000e+11,1251.16,12.0000,66.3600,253.3560,1.827,29.318,7.176
167,Pakistan,2015,271.0,4.50,5.50,4.70,2.710000e+11,Pakistan,PAK,South Asia,Lower middle income,2.710000e+11,1356.67,14.0000,66.5770,258.7000,3.566,29.124,7.107
168,Pakistan,2016,278.0,3.80,5.00,5.40,2.780000e+11,Pakistan,PAK,South Asia,Lower middle income,2.790000e+11,1368.45,12.3854,66.7700,264.1490,3.440,28.888,7.048
169,Pakistan,2017,314.0,4.00,5.00,5.40,3.140000e+11,Pakistan,PAK,South Asia,Lower middle income,3.050000e+11,1466.84,15.5100,66.9470,269.6870,3.181,28.599,6.994


#### Analytic Features

In [36]:
economic_worldbank_merged['gdp_difference'] = (
        economic_worldbank_merged['gdp_usd_economic'] -  economic_worldbank_merged['gdp_usd_worldbank']
)

economic_worldbank_merged['gdp_absolute_difference'] = (
    economic_worldbank_merged['gdp_difference'].abs()
)
economic_worldbank_merged[
    [
        "country",
        "year",
        "gdp_usd_economic",
        "gdp_usd_worldbank",
        "gdp_difference",
        "gdp_absolute_difference"
    ]
]



,country,year,gdp_usd_economic,gdp_usd_worldbank,gdp_difference,gdp_absolute_difference
0,United States,2010,1.500000e+13,1.500000e+13,0.000000e+00,0.000000e+00
1,United States,2011,1.550000e+13,1.550000e+13,0.000000e+00,0.000000e+00
2,United States,2012,1.600000e+13,1.620000e+13,-2.000000e+11,2.000000e+11
3,United States,2013,1.650000e+13,1.680000e+13,-3.000000e+11,3.000000e+11
4,United States,2014,1.700000e+13,1.750000e+13,-5.000000e+11,5.000000e+11
...,...,...,...,...,...,...
166,Pakistan,2014,2.470000e+11,2.440000e+11,3.000000e+09,3.000000e+09
167,Pakistan,2015,2.710000e+11,2.710000e+11,0.000000e+00,0.000000e+00
168,Pakistan,2016,2.780000e+11,2.790000e+11,-1.000000e+09,1.000000e+09
169,Pakistan,2017,3.140000e+11,3.050000e+11,9.000000e+09,9.000000e+09


In [37]:
economic_worldbank_merged['gdp_percentage_difference'] = (
    economic_worldbank_merged['gdp_absolute_difference'] /
    economic_worldbank_merged['gdp_usd_worldbank'] * 100
)

print(economic_worldbank_merged[
    [
        "country",
        "year",
        "gdp_usd_economic",
        "gdp_usd_worldbank",
        "gdp_difference",
        "gdp_absolute_difference",
        "gdp_percentage_difference"
    ]
].sort_values('gdp_percentage_difference', ascending=False))

           country  year  gdp_usd_economic  gdp_usd_worldbank  gdp_difference  \
143      Indonesia  2018      1.000000e+09       1.040000e+12   -1.039000e+12   
142      Indonesia  2017      1.000000e+09       1.020000e+12   -1.019000e+12   
147         Turkey  2013      1.000000e+09       9.510000e+11   -9.500000e+11   
148         Turkey  2014      1.000000e+09       9.340000e+11   -9.330000e+11   
141      Indonesia  2016      1.000000e+09       9.320000e+11   -9.310000e+11   
..             ...   ...               ...                ...             ...   
113         Brazil  2015      1.800000e+12       1.800000e+12    0.000000e+00   
165       Pakistan  2013      2.310000e+11       2.310000e+11    0.000000e+00   
167       Pakistan  2015      2.710000e+11       2.710000e+11    0.000000e+00   
1    United States  2011      1.550000e+13       1.550000e+13    0.000000e+00   
0    United States  2010      1.500000e+13       1.500000e+13    0.000000e+00   

     gdp_absolute_differenc

In [38]:
print(econom_ind_and_inflation_2010_2018.loc[
    econom_ind_and_inflation_2010_2018["country"].isin(["Indonesia", "Turkey"]),
    ["country", "year", "gdp_in_billion_usd", "gdb_usd_economic"]
].sort_values(["country", "year"]))

       country  year  gdp_in_billion_usd  gdb_usd_economic
240  Indonesia  2010               706.0      7.060000e+11
241  Indonesia  2011               846.0      8.460000e+11
242  Indonesia  2012                 1.0      1.000000e+09
243  Indonesia  2013                 1.0      1.000000e+09
244  Indonesia  2014                 1.0      1.000000e+09
245  Indonesia  2015                 1.0      1.000000e+09
246  Indonesia  2016                 1.0      1.000000e+09
247  Indonesia  2017                 1.0      1.000000e+09
248  Indonesia  2018                 1.0      1.000000e+09
256     Turkey  2010               774.0      7.740000e+11
257     Turkey  2011                 1.0      1.000000e+09
258     Turkey  2012                 1.0      1.000000e+09
259     Turkey  2013                 1.0      1.000000e+09
260     Turkey  2014                 1.0      1.000000e+09
261     Turkey  2015                 1.0      1.000000e+09
262     Turkey  2016                 1.0      1.000000e+

In [39]:
# World Bank GDP as the main GDP variable for analysis
economic_worldbank_merged['gdp_usd_final'] = economic_worldbank_merged['gdp_usd_worldbank']

After converting GDP from the Economic Indicators dataset from billion USD to USD, GDP values from both datasets were compared. Most values were reasonably similar, but some observations showed extremely large percentage differences. In particular, several GDP values for Indonesia and Turkey in the Economic Indicators dataset were recorded as 1.0 or 2.0 billion USD, while the World Bank values showed GDP levels close to hundreds of billions or around one trillion USD. This suggests a possible scale or data-entry issue in the Economic Indicators dataset.

Because of this inconsistency, the World Bank GDP variable was used as the main GDP measure for further analysis. The original Economic Indicators GDP variable was kept for comparison purposes, and GDP difference columns were retained to assess consistency between the two sources.

In [40]:
# Flag suspicious rows
economic_worldbank_merged['gdp_diff_status'] = np.where(
    economic_worldbank_merged['gdp_percentage_difference'] > 20,
    "Suspicious Difference",
    "Acceptable Difference"
)

economic_worldbank_merged[
    [
        "country",
        "year",
        "gdp_usd_economic",
        "gdp_usd_worldbank",
        "gdp_difference",
        "gdp_absolute_difference",
        "gdp_percentage_difference",
        'gdp_diff_status'
    ]
]


,country,year,gdp_usd_economic,gdp_usd_worldbank,gdp_difference,gdp_absolute_difference,gdp_percentage_difference,gdp_diff_status
0,United States,2010,1.500000e+13,1.500000e+13,0.000000e+00,0.000000e+00,0.000000,Acceptable Difference
1,United States,2011,1.550000e+13,1.550000e+13,0.000000e+00,0.000000e+00,0.000000,Acceptable Difference
2,United States,2012,1.600000e+13,1.620000e+13,-2.000000e+11,2.000000e+11,1.234568,Acceptable Difference
3,United States,2013,1.650000e+13,1.680000e+13,-3.000000e+11,3.000000e+11,1.785714,Acceptable Difference
4,United States,2014,1.700000e+13,1.750000e+13,-5.000000e+11,5.000000e+11,2.857143,Acceptable Difference
...,...,...,...,...,...,...,...,...
166,Pakistan,2014,2.470000e+11,2.440000e+11,3.000000e+09,3.000000e+09,1.229508,Acceptable Difference
167,Pakistan,2015,2.710000e+11,2.710000e+11,0.000000e+00,0.000000e+00,0.000000,Acceptable Difference
168,Pakistan,2016,2.780000e+11,2.790000e+11,-1.000000e+09,1.000000e+09,0.358423,Acceptable Difference
169,Pakistan,2017,3.140000e+11,3.050000e+11,9.000000e+09,9.000000e+09,2.950820,Acceptable Difference


In [41]:
suspicious_diffs = economic_worldbank_merged.loc[
    economic_worldbank_merged['gdp_diff_status'] == "Suspicious Difference",
]

print(f'Suspicious Differences Count: {len(suspicious_diffs)}')



Suspicious Differences Count: 40


In [42]:
suspicious_diffs[
    [
        "country",
        "year",
        "gdp_usd_economic",
        "gdp_usd_worldbank",
        "gdp_difference",
        "gdp_absolute_difference",
        "gdp_percentage_difference",
        'gdp_diff_status'
    ]
]

,country,year,gdp_usd_economic,gdp_usd_worldbank,gdp_difference,gdp_absolute_difference,gdp_percentage_difference,gdp_diff_status
12,China,2013,7.600000e+12,9.570000e+12,-1.970000e+12,1.970000e+12,20.585162,Suspicious Difference
13,China,2014,7.900000e+12,1.040000e+13,-2.500000e+12,2.500000e+12,24.038462,Suspicious Difference
14,China,2015,8.200000e+12,1.100000e+13,-2.800000e+12,2.800000e+12,25.454545,Suspicious Difference
15,China,2016,8.500000e+12,1.110000e+13,-2.600000e+12,2.600000e+12,23.423423,Suspicious Difference
16,China,2017,8.800000e+12,1.210000e+13,-3.300000e+12,3.300000e+12,27.272727,Suspicious Difference
17,China,2018,9.100000e+12,1.360000e+13,-4.500000e+12,4.500000e+12,33.088235,Suspicious Difference
22,Japan,2014,5.900000e+12,4.850000e+12,1.050000e+12,1.050000e+12,21.649485,Suspicious Difference
23,Japan,2015,6.000000e+12,4.390000e+12,1.610000e+12,1.610000e+12,36.674260,Suspicious Difference
24,Japan,2016,6.100000e+12,4.930000e+12,1.170000e+12,1.170000e+12,23.732252,Suspicious Difference
25,Japan,2017,6.200000e+12,4.860000e+12,1.340000e+12,1.340000e+12,27.572016,Suspicious Difference


Using a 20% threshold, 40 country-year observations were flagged as having suspicious GDP differences between the Economic Indicators dataset and the World Bank dataset. These observations were not removed automatically. Instead, they were flagged as potential scale or data-entry issues. For the final GDP-based analysis, the World Bank GDP variable was used as the main GDP measure because it was more consistent across countries and years.

In [43]:
print(economic_worldbank_merged['gdp_per_capita_usd'].describe())

count      171.000000
mean     24938.121117
std      19580.439447
min        781.154000
25%       8395.840000
50%      22087.000000
75%      42361.700000
max      68150.100000
Name: gdp_per_capita_usd, dtype: float64


In [44]:
# Set categories
q1, q3 = get_q1_q3(economic_worldbank_merged, 'gdp_per_capita_usd')

economic_worldbank_merged = add_categories(
    economic_worldbank_merged,
    'gdp_per_capita_usd',
    'gdp_per_capita_category',
    bins=[0, q1, q3, float("inf")],
    labels=['Low', 'Medium', 'High'],
    include_lowest=True,
)

economic_worldbank_merged[['country', 'year','gdp_per_capita_usd' ,'gdp_per_capita_category']].sort_values('gdp_per_capita_usd', ascending=False)

,country,year,gdp_per_capita_usd,gdp_per_capita_category
75,Australia,2013,68150.100,High
74,Australia,2012,68012.100,High
8,United States,2018,62641.000,High
73,Australia,2011,62517.800,High
76,Australia,2014,62510.800,High
...,...,...,...,...
162,Pakistan,2010,988.754,Low
129,Bangladesh,2013,981.840,Low
128,Bangladesh,2012,883.105,Low
127,Bangladesh,2011,861.758,Low


The quantile method was used to create data-driven categories. The 25th percentile and 75th percentile were used as thresholds. This means that the lowest 25% of GDP per capita observations were classified as Low, the middle 50% as Medium, and the highest 25% as High.

In [45]:
# Inflation Categories
zero_below = np.nextafter(0, -np.inf)

economic_worldbank_merged = add_categories(
    economic_worldbank_merged,
    'inflation_rate',
    'inflation_category',
    bins=[-float("inf"), zero_below, 3, 10,  float("inf")],
    labels=['Deflation','Low', 'Medium', 'High'],
    right=True
)

economic_worldbank_merged[
    [
        "country",
        "year",
        'inflation_rate',
        'inflation_category',
    ]
].sort_values('inflation_category', ascending=True)

,country,year,inflation_rate,inflation_category
24,Japan,2016,-0.50,Deflation
18,Japan,2010,-0.72,Deflation
123,Italy,2016,-0.10,Deflation
121,Italy,2014,-0.20,Deflation
0,United States,2010,1.64,Low
...,...,...,...,...
139,Indonesia,2014,492.00,High
138,Indonesia,2013,333.00,High
137,Indonesia,2012,84.00,High
36,India,2010,11.99,High


Inflation categories were defined using economic interpretation. Values below 0% were classified as deflation. Inflation from 0% to 3% was classified as low inflation, inflation above 3% and up to 10% as moderate inflation, and inflation above 10% as high inflation. These thresholds were chosen because low positive inflation is generally considered stable, while inflation above 10% indicates high inflation pressure.

In [46]:
economic_worldbank_merged['unemployment_rate'].describe()

count    171.000000
mean       6.482865
std        2.806592
min        2.400000
25%        4.450000
50%        5.600000
75%        8.150000
max       20.300000
Name: unemployment_rate, dtype: float64

In [47]:
# Unemployment category
q1_unemployment, q3_unemployment = get_q1_q3(economic_worldbank_merged, 'unemployment_rate')

economic_worldbank_merged = add_categories(
    df=economic_worldbank_merged,
    col='unemployment_rate',
    category_col_name='unemployment_category',
    bins=[0, q1_unemployment, q3_unemployment, float("inf")],
    labels=[
        'Low unemployment',
        'Medium unemployment',
        'High unemployment',
    ],
    include_lowest=True
)


economic_worldbank_merged[
    [
        "country",
        "year",
        'unemployment_rate',
        'unemployment_category',
    ]
].sort_values('unemployment_rate', ascending=False)


,country,year,unemployment_rate,unemployment_category
152,Turkey,2018,20.3,High unemployment
115,Brazil,2017,12.7,High unemployment
121,Italy,2014,12.7,High unemployment
116,Brazil,2018,12.3,High unemployment
120,Italy,2013,12.2,High unemployment
...,...,...,...,...
25,Japan,2017,3.0,Low unemployment
155,Malaysia,2012,3.0,Low unemployment
156,Malaysia,2013,2.9,Low unemployment
157,Malaysia,2014,2.8,Low unemployment


Unemployment categories were created using the distribution of unemployment rates in the merged dataset. The 25th percentile and 75th percentile were used as thresholds. Observations below the 25th percentile were classified as Low unemployment, observations between the 25th and 75th percentile as Medium unemployment, and observations above the 75th percentile as High unemployment. This data-driven approach was chosen because unemployment levels differ across countries and the categories should reflect the distribution within this analytical sample.

In [48]:
economic_worldbank_merged[['country', 'individuals_using_the_internet_of_population']]

,country,individuals_using_the_internet_of_population
0,United States,71.6900
1,United States,69.7295
2,United States,74.7000
3,United States,71.4000
4,United States,73.0000
...,...,...
166,Pakistan,12.0000
167,Pakistan,14.0000
168,Pakistan,12.3854
169,Pakistan,15.5100


In [49]:
# Internet Usage categories
q1, q3 = get_q1_q3(economic_worldbank_merged, 'individuals_using_the_internet_of_population')

economic_worldbank_merged = add_categories(
    economic_worldbank_merged,
    col='individuals_using_the_internet_of_population',
    category_col_name='internet_usage_category',
    bins=[0, q1, q3, float("inf")],
    labels=['Low', 'Medium', 'High'],
    include_lowest=True
)

economic_worldbank_merged[['country', 'individuals_using_the_internet_of_population', 'internet_usage_category']]


,country,individuals_using_the_internet_of_population,internet_usage_category
0,United States,71.6900,Medium
1,United States,69.7295,Medium
2,United States,74.7000,Medium
3,United States,71.4000,Medium
4,United States,73.0000,Medium
...,...,...,...
166,Pakistan,12.0000,Low
167,Pakistan,14.0000,Low
168,Pakistan,12.3854,Low
169,Pakistan,15.5100,Low


In [50]:
# Turning all Category type columns into string type
for col in economic_worldbank_merged.columns:
    if 'category' in col:
        economic_worldbank_merged[col] = economic_worldbank_merged[col].astype(str)

In [51]:
# Handle NaN values in Internet Usage Category
economic_worldbank_merged['internet_usage_category'] = (
    economic_worldbank_merged['internet_usage_category'].fillna('Unknown')
)

economic_worldbank_merged[['country', 'individuals_using_the_internet_of_population', 'internet_usage_category']]

,country,individuals_using_the_internet_of_population,internet_usage_category
0,United States,71.6900,Medium
1,United States,69.7295,Medium
2,United States,74.7000,Medium
3,United States,71.4000,Medium
4,United States,73.0000,Medium
...,...,...,...
166,Pakistan,12.0000,Low
167,Pakistan,14.0000,Low
168,Pakistan,12.3854,Low
169,Pakistan,15.5100,Low


In [52]:
#National Population Growth
economic_worldbank_merged['population_growth'] = (
    economic_worldbank_merged['birth_rate_crude_per_1_000_people'] -
    economic_worldbank_merged['death_rate_crude_per_1_000_people']
)

economic_worldbank_merged[
    [
        'country',
        'year',
        'birth_rate_crude_per_1_000_people',
        'death_rate_crude_per_1_000_people',
        'population_growth'
    ]
].sort_values('population_growth', ascending=False)

,country,year,birth_rate_crude_per_1_000_people,death_rate_crude_per_1_000_people,population_growth
162,Pakistan,2010,30.082,7.568,22.514
163,Pakistan,2011,29.860,7.454,22.406
164,Pakistan,2012,29.667,7.349,22.318
165,Pakistan,2013,29.491,7.257,22.234
166,Pakistan,2014,29.318,7.176,22.142
...,...,...,...,...,...
134,Bangladesh,2018,NaN,NaN,NaN
143,Indonesia,2018,NaN,NaN,NaN
152,Turkey,2018,NaN,NaN,NaN
161,Malaysia,2018,NaN,NaN,NaN


In [53]:
# Average GDP, inflation, unemployment and economic growth per country
avg_economy_by_country = (
    economic_worldbank_merged
    .groupby('country')
    .agg(
        avg_gdp=('gdp_usd_final', 'mean'),
        avg_inflation=('inflation_rate', 'mean'),
        avg_unemployment=('unemployment_rate', 'mean'),
        avg_economic_growth=('economic_growth', 'mean'),
    )
    .reset_index()
)

avg_economy_by_country.sort_values('avg_inflation', ascending=True)

,country,avg_gdp,avg_inflation,avg_unemployment,avg_economic_growth
10,Japan,5.246667e+12,0.433333,3.911111,1.411111
9,Italy,2.053333e+12,1.122222,10.866667,0.177778
5,France,2.680000e+12,1.155556,9.888889,1.022222
6,Germany,3.660000e+12,1.355556,5.277778,3.388889
3,Canada,1.700000e+12,1.711111,6.933333,2.177778
18,United States,1.754444e+13,1.771111,6.668889,2.467778
11,"Korea, Rep.",1.352222e+12,1.866667,3.488889,3.300000
0,Australia,1.385556e+12,2.111111,5.511111,2.877778
12,Malaysia,3.101111e+11,2.200000,3.122222,5.377778
17,United Kingdom,2.730000e+12,2.277778,6.188889,1.755556


In [54]:
# Average GDP, GDP per capita, unemployment, and life expectancy by region
avg_economy_by_region = (
    economic_worldbank_merged
    .groupby('region')
    .agg(
        avg_gdp=('gdp_usd_final', 'mean'),
        avg_gdp_per_capita=('gdp_per_capita_usd', 'mean'),
        avg_unemployment=('unemployment_rate', 'mean'),
        avg_life_expectancy=('life_expectancy_at_birth_years', 'mean'),
    )
    .reset_index()
)

avg_economy_by_region

,region,avg_gdp,avg_gdp_per_capita,avg_unemployment,avg_life_expectancy
0,East Asia & Pacific,3.200259e+12,24696.483148,4.260185,77.990865
1,Europe & Central Asia,2.293278e+12,30862.672963,8.027778,78.843573
2,Latin America & Caribbean,2.194444e+12,10849.255556,8.211111,74.582125
3,Middle East & North Africa,6.897778e+11,22454.477778,10.500000,74.435125
4,North America,9.622222e+12,51547.772222,6.801111,80.216225
5,South Asia,8.480741e+11,1356.695593,5.711111,68.414792


In [55]:
# Average inflation and unemployment by income group
avg_inflation_unemployment_by_income_group = (
    economic_worldbank_merged
    .groupby('income_group')
    .agg(
        avg_inflation=('inflation_rate', 'mean'),
        avg_unemployment=('unemployment_rate', 'mean'),
    )
    .reset_index()
)

avg_inflation_unemployment_by_income_group

,income_group,avg_inflation,avg_unemployment
0,High income: OECD,1.533827,6.526173
1,High income: nonOECD,6.516667,8.083333
2,Lower middle income,34.778056,5.658333
3,Upper middle income,137.553611,6.409722


In [56]:
economic_worldbank_merged.inflation_category.unique()

<StringArray>
['Low', 'Medium', 'Deflation', 'High']
Length: 4, dtype: str

In [57]:
# Countries with the highest and lowest average inflation
highest_inflation_country = highest(avg_economy_by_country, 'avg_inflation')
lowest_inflation_country = lowest(avg_economy_by_country, 'avg_inflation')

print(highest_print('Inflation', highest_inflation_country['country'], highest_inflation_country['avg_inflation']))
print()
print(lowest_print('Inflation', lowest_inflation_country['country'], lowest_inflation_country['avg_inflation']))

Highest average inflation:
- Turkey
- Inflation: 539.27

Lowest average inflation:
- Japan
- Inflation: 0.43


In [58]:
# Countries with the highest and lowest average unemployment
highest_avg_unemployment_country =highest(avg_economy_by_country, 'avg_unemployment')
lowest_avg_unemployment_country  = lowest(avg_economy_by_country, 'avg_unemployment')

print(highest_print('Unemployment', highest_inflation_country['country'], highest_inflation_country['avg_unemployment']))
print()
print(lowest_print('Unemployment', lowest_inflation_country['country'], lowest_inflation_country['avg_unemployment']))

Highest average unemployment:
- Turkey
- Unemployment: 10.28

Lowest average unemployment:
- Japan
- Unemployment: 3.91


In [59]:
# Countries with the highest and lowest GDP per capita
highest_gdp_country= highest(avg_economy_by_country, 'avg_gdp')
lowest_gdp_country = lowest(avg_economy_by_country, 'avg_gdp')

print(highest_print('GDP', highest_gdp_country['country'], highest_gdp_country['avg_gdp']))
print()
print(lowest_print('GDP', lowest_gdp_country['country'], lowest_gdp_country['avg_gdp']))


Highest average gdp:
- United States
- GDP: 17544444444444.45

Lowest average gdp:
- Bangladesh
- GDP: 182222222222.22


In [60]:
# Average internet usage by region and income group
avg_internet_usage_income_group = (
    economic_worldbank_merged
    .groupby('income_group')
    .agg(
        avg_internet_usage=('individuals_using_the_internet_of_population', 'mean'),
    )
    .reset_index()
)

avg_internet_usage_income_group

,income_group,avg_internet_usage
0,High income: OECD,81.477577
1,High income: nonOECD,65.666300
2,Lower middle income,15.209361
3,Upper middle income,55.344300


In [61]:
# Relationship between income group and life expectancy.
life_expectancy_by_income_group = (
    economic_worldbank_merged
    .groupby('income_group')
    .agg(
        avg_life_expectancy=('life_expectancy_at_birth_years', 'mean'),
    )
    .reset_index()
)

life_expectancy_by_income_group

,income_group,avg_life_expectancy
0,High income: OECD,81.554572
1,High income: nonOECD,72.522150
2,Lower middle income,68.884781
3,Upper middle income,75.281531


In [62]:
# Average birth rate and death rate by region.
avg_birth_rate_by_region = (
    economic_worldbank_merged
    .groupby('region')
    .agg(
        avg_birth_rate=('birth_rate_crude_per_1_000_people', 'mean'),
    )
    .reset_index()
)

avg_birth_rate_by_region

,region,avg_birth_rate
0,East Asia & Pacific,13.149958
1,Europe & Central Asia,11.971938
2,Latin America & Caribbean,14.708375
3,Middle East & North Africa,20.237250
4,North America,11.618750
5,South Asia,22.827250


In [63]:
# Countries with the strongest economic growth over time
economy_growth_top5 = avg_economy_by_country.sort_values('avg_economic_growth', ascending=False).head(5)[['country','avg_economic_growth']]
economy_growth_top5

,country,avg_economic_growth
16,Turkey,10.011111
4,China,7.722222
7,India,7.277778
1,Bangladesh,6.688889
8,Indonesia,5.911111
